File: projection.py
"""Pure projection engine — no plotting, no UI, fully JSON-serializable.

Used by both the Streamlit UI (app.py) and the FastAPI backend (api.py).
"""
from __future__ import annotations

from dataclasses import dataclass, field, asdict
from typing import Optional

import numpy as np


# ---------------------------------------------------------------------------
# Constants & asset metadata
# ---------------------------------------------------------------------------

SUPPORTED_ASSETS: list[str] = [
    "XAUUSD", "XAGUSD", "BTC", "ETH", "XU100", "XU30", "NASDAQ", "S&P",
]


# ---------------------------------------------------------------------------
# Inputs
# ---------------------------------------------------------------------------

@dataclass
class AssetParams:
    """Per-asset macro params (sourced from yfinance via stock_market_helper)."""
    average_growth: float  # yearly % (e.g. 12.3 for 12.3%/yr)
    current_price: float   # USD


@dataclass
class ProjectionInput:
    # Loan
    years: int = 10
    interest_rate: float = 2.74                # monthly %, e.g. 2.74
    initial_noncredit_amount_tl: float = 350_000
    value_of_house_tl: float = 2_100_000

    # Macro
    start_dollar_tl: float = 45.0
    dollar_growth_rate: float = 35.0           # yearly %, nominal
    turkey_inflation: float = 25.0             # yearly %
    usa_inflation: float = 3.0                 # yearly %

    # Rent
    initial_monthly_rent_tl: float = 20_000.0

    # Salary
    salary_currency: str = "USD"               # one of {EUR, USD, TL}
    start_salary_base: float = 0.0             # annual figure
    salary_growth: float = 0.0                 # yearly %
    euro_dollar_rate: float = 1.15
    months_to_increase: int = 12

    # Toggles
    use_months: bool = True
    deflate_dollar_by_us_inflation: bool = False  # subtract USA inflation from dollar growth
    generate_random: bool = False
    project_initial_money_with_asset: bool = False

    # Assets — symbol -> params; presence means "include in projection"
    assets: dict[str, AssetParams] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# Core financial primitives
# ---------------------------------------------------------------------------

def compound_series(start: float, rate: float, steps: int) -> np.ndarray:
    """Return array of length steps+1: [start, start*(1+r), start*(1+r)^2, ...]."""
    return start * np.power(1.0 + rate, np.arange(steps + 1))


def amortized_monthly_payment(principal: float, monthly_rate: float, n_months: int) -> float:
    """Standard mortgage amortization formula. Returns the fixed monthly payment.

    M = P * r(1+r)^n / ((1+r)^n - 1)

    Zero-interest edge case: M = P / n.
    """
    if n_months <= 0:
        raise ValueError("n_months must be positive")
    if monthly_rate == 0:
        return principal / n_months
    factor = (1 + monthly_rate) ** n_months
    return principal * monthly_rate * factor / (factor - 1)


def random_yearly_path(initial: float, years: int, target_avg_pct: float) -> tuple[np.ndarray, np.ndarray]:
    """Generate a yearly path with random jitter that scales to hit the target average."""
    target = target_avg_pct / 100.0
    raw = np.random.uniform(target * 0.7, target * 1.3, years)
    adjusted = raw * (target / np.mean(raw)) if np.mean(raw) != 0 else raw
    path = [initial]
    for r in adjusted:
        path.append(path[-1] * (1 + r))
    return np.array(path), adjusted


def random_monthly_path(initial: float, years: int, target_avg_pct: float) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Generate yearly + monthly paths where monthly multipliers normalize to yearly target."""
    yearly_path, adjusted_yearly = random_yearly_path(initial, years, target_avg_pct)
    monthly = [initial]
    for y_idx in range(years):
        year_multiplier = 1 + adjusted_yearly[y_idx]
        raw_monthly = np.random.uniform(0.98, 1.05, 12)
        # normalize so product = year_multiplier
        normalized = raw_monthly * (year_multiplier ** (1 / 12)) / (np.prod(raw_monthly) ** (1 / 12))
        for m in range(12):
            monthly.append(monthly[-1] * normalized[m])
    return yearly_path, adjusted_yearly, np.array(monthly)


# ---------------------------------------------------------------------------
# Projection
# ---------------------------------------------------------------------------

def _effective_dollar_growth(inp: ProjectionInput) -> tuple[float, float]:
    """Return (annual_decimal, monthly_decimal) growth rate after optional US-inflation deflation."""
    annual_pct = inp.dollar_growth_rate
    if inp.deflate_dollar_by_us_inflation:
        annual_pct -= inp.usa_inflation
    annual = annual_pct / 100.0
    monthly = (1 + annual) ** (1 / 12) - 1
    return annual, monthly


def _build_salary_series_yearly(inp: ProjectionInput, dollar_rates_annual: np.ndarray) -> np.ndarray:
    """Yearly salary series, converted to USD."""
    salaries = [inp.start_salary_base]
    for _ in range(1, len(dollar_rates_annual)):
        salaries.append(salaries[-1] * (1 + inp.salary_growth / 100))
    arr = np.array(salaries)
    if inp.salary_currency == "EUR":
        return arr * inp.euro_dollar_rate
    if inp.salary_currency == "USD":
        return arr
    if inp.salary_currency == "TL":
        return arr / dollar_rates_annual
    raise ValueError(f"unknown salary currency: {inp.salary_currency}")


def _monthly_from_yearly_salary(yearly_arr: np.ndarray, months_to_increase: int,
                                dollar_rates_monthly: np.ndarray,
                                euro_dollar_rate: float, salary_currency: str) -> np.ndarray:
    """Spread a yearly salary array into a monthly array honouring months_to_increase offset."""
    monthly: list[float] = []
    monthly.extend([yearly_arr[0] / 12] * months_to_increase)
    for y in yearly_arr[1:-1]:
        monthly.extend([y / 12] * 12)
    monthly.extend([yearly_arr[-1] / 12] * (13 - months_to_increase))

    arr = np.array(monthly)
    if salary_currency == "EUR":
        return arr * euro_dollar_rate
    if salary_currency == "USD":
        return arr
    if salary_currency == "TL":
        # truncate / extend to dollar_rates_monthly length to match shape
        m = min(len(arr), len(dollar_rates_monthly))
        return arr[:m] / dollar_rates_monthly[:m]
    raise ValueError(f"unknown salary currency: {salary_currency}")


def _project_asset(asset: AssetParams, inp: ProjectionInput) -> tuple[np.ndarray, np.ndarray]:
    """Return (yearly_path, monthly_path) for one asset given its growth rate."""
    start_price = (
        inp.initial_noncredit_amount_tl / inp.start_dollar_tl
        if inp.project_initial_money_with_asset
        else asset.current_price
    )
    growth_pct = asset.average_growth
    if inp.generate_random:
        yearly, _, monthly = random_monthly_path(start_price, inp.years, growth_pct)
    else:
        yearly = compound_series(start_price, growth_pct / 100, inp.years)
        monthly_rate = (1 + growth_pct / 100) ** (1 / 12) - 1
        monthly = compound_series(start_price, monthly_rate, inp.years * 12)
    return yearly, monthly


@dataclass
class ProjectionResult:
    # Metadata
    monthly_tl_payment: float
    annual_tl_payment: float
    total_paid_tl: float                       # monthly_tl_payment * years * 12
    loan_amount_tl: float
    initial_noncredit_amount_usd: float
    value_of_house_usd: float

    # Effective rates used
    effective_dollar_growth_annual: float      # decimal (e.g. 0.32)
    effective_dollar_growth_monthly: float
    effective_turkey_inflation_annual: float
    effective_turkey_inflation_monthly: float

    # Yearly series (length years+1)
    years_axis: list[int]
    dollar_rates_annual: list[float]
    annual_payment_usd: list[float]            # length years
    cumulative_payment_usd_annual: list[float] # length years
    total_credit_amount_usd_annual: list[float]  # cumulative + initial_noncredit_usd, length years+1
    value_of_house_usd_yearly: list[float]
    rent_price_usd_yearly: list[float]
    cumulative_rent_price_usd_yearly: list[float]
    house_plus_rent_yearly: list[float]
    salaries_usd_yearly: Optional[list[float]] = None
    payment_salary_ratio_yearly: Optional[list[float]] = None
    payment_minus_rent_over_salary_yearly: Optional[list[float]] = None

    # Monthly series (length years*12 + 1)
    months_axis: list[int] = field(default_factory=list)
    dollar_rates_monthly: list[float] = field(default_factory=list)
    monthly_payment_usd: list[float] = field(default_factory=list)
    cumulative_payment_usd_monthly: list[float] = field(default_factory=list)
    total_credit_amount_usd_monthly: list[float] = field(default_factory=list)
    value_of_house_usd_monthly: list[float] = field(default_factory=list)
    rent_price_usd_monthly: list[float] = field(default_factory=list)
    cumulative_rent_price_usd_monthly: list[float] = field(default_factory=list)
    house_plus_rent_monthly: list[float] = field(default_factory=list)
    salaries_usd_monthly: Optional[list[float]] = None
    payment_salary_ratio_monthly: Optional[list[float]] = None
    payment_minus_rent_over_salary_monthly: Optional[list[float]] = None

    # Asset projections: {symbol: {"yearly": [...], "monthly": [...], "average_growth": x, "current_price": y}}
    asset_projections: dict[str, dict] = field(default_factory=dict)

    def to_dict(self) -> dict:
        return asdict(self)


def run_projection(inp: ProjectionInput) -> ProjectionResult:
    n_months = inp.years * 12

    # --- Loan
    loan = inp.value_of_house_tl - inp.initial_noncredit_amount_tl
    monthly_tl_payment = amortized_monthly_payment(loan, inp.interest_rate / 100, n_months)
    annual_tl_payment = monthly_tl_payment * 12

    # --- Effective rates
    dollar_growth_annual, dollar_growth_monthly = _effective_dollar_growth(inp)
    turkey_inflation_annual = inp.turkey_inflation / 100
    turkey_inflation_monthly = (1 + turkey_inflation_annual) ** (1 / 12) - 1

    # --- Dollar rate paths
    if inp.generate_random:
        dollar_rates_annual, _, dollar_rates_monthly = random_monthly_path(
            inp.start_dollar_tl, inp.years, dollar_growth_annual * 100,
        )
    else:
        dollar_rates_annual = compound_series(inp.start_dollar_tl, dollar_growth_annual, inp.years)
        dollar_rates_monthly = compound_series(inp.start_dollar_tl, dollar_growth_monthly, n_months)

    # --- Payment in USD
    annual_payment_usd = np.full(inp.years, annual_tl_payment) / dollar_rates_annual[1:]
    monthly_payment_usd = np.full(n_months, monthly_tl_payment) / dollar_rates_monthly[1:]
    cum_payment_usd_annual = np.cumsum(annual_payment_usd)
    cum_payment_usd_monthly = np.cumsum(monthly_payment_usd)

    initial_noncredit_usd = inp.initial_noncredit_amount_tl / inp.start_dollar_tl
    total_credit_usd_annual = np.insert(cum_payment_usd_annual, 0, 0) + initial_noncredit_usd
    total_credit_usd_monthly = np.insert(cum_payment_usd_monthly, 0, 0) + initial_noncredit_usd

    # --- House value
    if inp.generate_random:
        house_tl_yearly, _, house_tl_monthly = random_monthly_path(
            inp.value_of_house_tl, inp.years, inp.turkey_inflation,
        )
    else:
        house_tl_yearly = compound_series(inp.value_of_house_tl, turkey_inflation_annual, inp.years)
        house_tl_monthly = compound_series(inp.value_of_house_tl, turkey_inflation_monthly, n_months)
    house_usd_yearly = house_tl_yearly / dollar_rates_annual
    house_usd_monthly = house_tl_monthly / dollar_rates_monthly

    # --- Rent
    annual_rent_tl_start = inp.initial_monthly_rent_tl * 12
    if inp.generate_random:
        rent_tl_yearly, _, rent_tl_monthly_year_grouped = random_monthly_path(
            annual_rent_tl_start, inp.years, inp.turkey_inflation,
        )
        # Monthly rent: divide each year's annual rent by 12, repeat 12 times
        rent_tl_monthly: list[float] = []
        for y in rent_tl_yearly[:-1]:
            rent_tl_monthly.extend([y / 12] * 12)
        rent_tl_monthly.append(rent_tl_yearly[-1] / 12)
        rent_tl_monthly_arr = np.array(rent_tl_monthly)
    else:
        rent_tl_yearly = compound_series(annual_rent_tl_start, turkey_inflation_annual, inp.years)
        rent_tl_monthly_list: list[float] = []
        for y in rent_tl_yearly[:-1]:
            rent_tl_monthly_list.extend([y / 12] * 12)
        rent_tl_monthly_list.append(rent_tl_yearly[-1] / 12)
        rent_tl_monthly_arr = np.array(rent_tl_monthly_list)
    rent_usd_yearly = rent_tl_yearly / dollar_rates_annual
    rent_usd_monthly = rent_tl_monthly_arr / dollar_rates_monthly
    cum_rent_usd_yearly = np.cumsum(rent_usd_yearly)
    cum_rent_usd_monthly = np.cumsum(rent_usd_monthly)
    house_plus_rent_yearly = house_usd_yearly + cum_rent_usd_yearly
    house_plus_rent_monthly = house_usd_monthly + cum_rent_usd_monthly

    # --- Salary
    salaries_usd_yearly_arr = None
    salaries_usd_monthly_arr = None
    pay_sal_ratio_yearly = None
    pay_sal_ratio_monthly = None
    pay_minus_rent_yearly = None
    pay_minus_rent_monthly = None

    if inp.start_salary_base > 0:
        if inp.generate_random:
            same_currency_yearly, _ = random_yearly_path(
                inp.start_salary_base, inp.years, inp.salary_growth,
            )
        else:
            same_currency_yearly = compound_series(
                inp.start_salary_base, inp.salary_growth / 100, inp.years,
            )

        # Convert yearly to USD (length years+1)
        if inp.salary_currency == "EUR":
            salaries_usd_yearly_arr = same_currency_yearly * inp.euro_dollar_rate
        elif inp.salary_currency == "USD":
            salaries_usd_yearly_arr = same_currency_yearly.copy()
        elif inp.salary_currency == "TL":
            salaries_usd_yearly_arr = same_currency_yearly / dollar_rates_annual

        # Monthly
        salaries_usd_monthly_arr = _monthly_from_yearly_salary(
            same_currency_yearly, inp.months_to_increase,
            dollar_rates_monthly, inp.euro_dollar_rate, inp.salary_currency,
        )

        # Ratios (skip first salary entry which is t=0)
        sal_y = salaries_usd_yearly_arr[1:]
        pay_sal_ratio_yearly = annual_payment_usd / sal_y
        pay_minus_rent_yearly = (annual_payment_usd - rent_usd_yearly[1:]) / sal_y

        sal_m = salaries_usd_monthly_arr[:n_months]
        # Align lengths
        pay_sal_ratio_monthly = monthly_payment_usd[:len(sal_m)] / sal_m
        pay_minus_rent_monthly = (
            (monthly_payment_usd[:len(sal_m)] - rent_usd_monthly[1:len(sal_m) + 1])
            / sal_m
        )

    # --- Assets
    asset_projections: dict[str, dict] = {}
    for symbol, params in inp.assets.items():
        yearly_path, monthly_path = _project_asset(params, inp)
        asset_projections[symbol] = {
            "yearly": yearly_path.tolist(),
            "monthly": monthly_path.tolist(),
            "average_growth": params.average_growth,
            "current_price": params.current_price,
        }

    return ProjectionResult(
        monthly_tl_payment=float(monthly_tl_payment),
        annual_tl_payment=float(annual_tl_payment),
        total_paid_tl=float(monthly_tl_payment * n_months),
        loan_amount_tl=float(loan),
        initial_noncredit_amount_usd=float(initial_noncredit_usd),
        value_of_house_usd=float(inp.value_of_house_tl / inp.start_dollar_tl),
        effective_dollar_growth_annual=float(dollar_growth_annual),
        effective_dollar_growth_monthly=float(dollar_growth_monthly),
        effective_turkey_inflation_annual=float(turkey_inflation_annual),
        effective_turkey_inflation_monthly=float(turkey_inflation_monthly),
        years_axis=list(range(inp.years + 1)),
        months_axis=list(range(n_months + 1)),
        dollar_rates_annual=dollar_rates_annual.tolist(),
        dollar_rates_monthly=dollar_rates_monthly.tolist(),
        annual_payment_usd=annual_payment_usd.tolist(),
        monthly_payment_usd=monthly_payment_usd.tolist(),
        cumulative_payment_usd_annual=cum_payment_usd_annual.tolist(),
        cumulative_payment_usd_monthly=cum_payment_usd_monthly.tolist(),
        total_credit_amount_usd_annual=total_credit_usd_annual.tolist(),
        total_credit_amount_usd_monthly=total_credit_usd_monthly.tolist(),
        value_of_house_usd_yearly=house_usd_yearly.tolist(),
        value_of_house_usd_monthly=house_usd_monthly.tolist(),
        rent_price_usd_yearly=rent_usd_yearly.tolist(),
        rent_price_usd_monthly=rent_usd_monthly.tolist(),
        cumulative_rent_price_usd_yearly=cum_rent_usd_yearly.tolist(),
        cumulative_rent_price_usd_monthly=cum_rent_usd_monthly.tolist(),
        house_plus_rent_yearly=house_plus_rent_yearly.tolist(),
        house_plus_rent_monthly=house_plus_rent_monthly.tolist(),
        salaries_usd_yearly=salaries_usd_yearly_arr.tolist() if salaries_usd_yearly_arr is not None else None,
        salaries_usd_monthly=salaries_usd_monthly_arr.tolist() if salaries_usd_monthly_arr is not None else None,
        payment_salary_ratio_yearly=pay_sal_ratio_yearly.tolist() if pay_sal_ratio_yearly is not None else None,
        payment_salary_ratio_monthly=pay_sal_ratio_monthly.tolist() if pay_sal_ratio_monthly is not None else None,
        payment_minus_rent_over_salary_yearly=pay_minus_rent_yearly.tolist() if pay_minus_rent_yearly is not None else None,
        payment_minus_rent_over_salary_monthly=pay_minus_rent_monthly.tolist() if pay_minus_rent_monthly is not None else None,
        asset_projections=asset_projections,
    )

File: frontend/projection.js
// =============================================================
// Pure projection engine — browser port of projection.py
//
// Same outputs as ProjectionResult.to_dict() so the rest of app.js
// (SERIES_MAP, asset_projections, etc.) works unchanged.
// =============================================================

const SUPPORTED_ASSETS = [
  "XAUUSD", "XAGUSD", "BTC", "ETH", "XU100", "XU30", "NASDAQ", "S&P",
];

// ---------------- primitives ----------------

function compoundSeries(start, rate, steps) {
  const out = new Array(steps + 1);
  out[0] = start;
  for (let i = 1; i <= steps; i++) out[i] = out[i - 1] * (1 + rate);
  return out;
}

function amortizedMonthlyPayment(principal, monthlyRate, nMonths) {
  if (nMonths <= 0) throw new Error("n_months must be positive");
  if (monthlyRate === 0) return principal / nMonths;
  const factor = Math.pow(1 + monthlyRate, nMonths);
  return (principal * monthlyRate * factor) / (factor - 1);
}

function uniform(a, b) { return a + Math.random() * (b - a); }

function randomYearlyPath(initial, years, targetAvgPct) {
  const target = targetAvgPct / 100;
  const raw = Array.from({ length: years }, () => uniform(target * 0.7, target * 1.3));
  const mean = raw.reduce((s, x) => s + x, 0) / years;
  const adjusted = mean !== 0 ? raw.map(r => r * (target / mean)) : raw;
  const path = [initial];
  for (const r of adjusted) path.push(path[path.length - 1] * (1 + r));
  return { path, adjusted };
}

function randomMonthlyPath(initial, years, targetAvgPct) {
  const { path: yearlyPath, adjusted } = randomYearlyPath(initial, years, targetAvgPct);
  const monthly = [initial];
  for (let y = 0; y < years; y++) {
    const yearMul = 1 + adjusted[y];
    const rawM = Array.from({ length: 12 }, () => uniform(0.98, 1.05));
    const productPow = Math.pow(rawM.reduce((p, x) => p * x, 1), 1 / 12);
    const yearMulPow = Math.pow(yearMul, 1 / 12);
    const norm = rawM.map(x => (x * yearMulPow) / productPow);
    for (let m = 0; m < 12; m++) monthly.push(monthly[monthly.length - 1] * norm[m]);
  }
  return { yearlyPath, adjusted, monthly };
}

// element-wise array helpers (numpy-ish)
const div = (a, b) => a.map((x, i) => x / b[i]);
const sub = (a, b) => a.map((x, i) => x - b[i]);
const add = (a, b) => a.map((x, i) => x + b[i]);
const scale = (a, s) => a.map(x => x * s);
function cumsum(a) {
  const out = new Array(a.length);
  let s = 0;
  for (let i = 0; i < a.length; i++) { s += a[i]; out[i] = s; }
  return out;
}

// ---------------- intermediate helpers ----------------

function effectiveDollarGrowth(inp) {
  let annualPct = inp.dollar_growth_rate;
  if (inp.deflate_dollar_by_us_inflation) annualPct -= inp.usa_inflation;
  const annual = annualPct / 100;
  const monthly = Math.pow(1 + annual, 1 / 12) - 1;
  return { annual, monthly };
}

function monthlyFromYearlySalary(yearlyArr, monthsToIncrease, dollarRatesMonthly, euroDollarRate, salaryCurrency) {
  const monthly = [];
  for (let i = 0; i < monthsToIncrease; i++) monthly.push(yearlyArr[0] / 12);
  for (let i = 1; i < yearlyArr.length - 1; i++) {
    for (let j = 0; j < 12; j++) monthly.push(yearlyArr[i] / 12);
  }
  for (let i = 0; i < 13 - monthsToIncrease; i++) monthly.push(yearlyArr[yearlyArr.length - 1] / 12);

  if (salaryCurrency === "EUR") return monthly.map(x => x * euroDollarRate);
  if (salaryCurrency === "USD") return monthly;
  if (salaryCurrency === "TL") {
    const m = Math.min(monthly.length, dollarRatesMonthly.length);
    return monthly.slice(0, m).map((x, i) => x / dollarRatesMonthly[i]);
  }
  throw new Error(`unknown salary currency: ${salaryCurrency}`);
}

function projectAsset(asset, inp) {
  const startPrice = inp.project_initial_money_with_asset
    ? inp.initial_noncredit_amount_tl / inp.start_dollar_tl
    : asset.current_price;
  const growthPct = asset.average_growth;
  if (inp.generate_random) {
    const { yearlyPath, monthly } = randomMonthlyPath(startPrice, inp.years, growthPct);
    return { yearly: yearlyPath, monthly };
  }
  const yearly = compoundSeries(startPrice, growthPct / 100, inp.years);
  const monthlyRate = Math.pow(1 + growthPct / 100, 1 / 12) - 1;
  const monthly = compoundSeries(startPrice, monthlyRate, inp.years * 12);
  return { yearly, monthly };
}

// ---------------- main entry ----------------

function runProjection(inp) {
  const nMonths = inp.years * 12;

  // Loan
  const loan = inp.value_of_house_tl - inp.initial_noncredit_amount_tl;
  const monthlyTlPayment = amortizedMonthlyPayment(loan, inp.interest_rate / 100, nMonths);
  const annualTlPayment = monthlyTlPayment * 12;

  // Effective rates
  const { annual: dollarGrowthAnnual, monthly: dollarGrowthMonthly } = effectiveDollarGrowth(inp);
  const turkeyInflationAnnual = inp.turkey_inflation / 100;
  const turkeyInflationMonthly = Math.pow(1 + turkeyInflationAnnual, 1 / 12) - 1;

  // Dollar rate paths
  let dollarRatesAnnual, dollarRatesMonthly;
  if (inp.generate_random) {
    const r = randomMonthlyPath(inp.start_dollar_tl, inp.years, dollarGrowthAnnual * 100);
    dollarRatesAnnual = r.yearlyPath;
    dollarRatesMonthly = r.monthly;
  } else {
    dollarRatesAnnual = compoundSeries(inp.start_dollar_tl, dollarGrowthAnnual, inp.years);
    dollarRatesMonthly = compoundSeries(inp.start_dollar_tl, dollarGrowthMonthly, nMonths);
  }

  // Payment in USD
  const annualPaymentUsd = dollarRatesAnnual.slice(1).map(r => annualTlPayment / r);
  const monthlyPaymentUsd = dollarRatesMonthly.slice(1).map(r => monthlyTlPayment / r);
  const cumPaymentUsdAnnual = cumsum(annualPaymentUsd);
  const cumPaymentUsdMonthly = cumsum(monthlyPaymentUsd);

  const initialNoncreditUsd = inp.initial_noncredit_amount_tl / inp.start_dollar_tl;
  const totalCreditUsdAnnual = [initialNoncreditUsd, ...cumPaymentUsdAnnual.map(x => x + initialNoncreditUsd)];
  const totalCreditUsdMonthly = [initialNoncreditUsd, ...cumPaymentUsdMonthly.map(x => x + initialNoncreditUsd)];

  // House value
  let houseTlYearly, houseTlMonthly;
  if (inp.generate_random) {
    const r = randomMonthlyPath(inp.value_of_house_tl, inp.years, inp.turkey_inflation);
    houseTlYearly = r.yearlyPath;
    houseTlMonthly = r.monthly;
  } else {
    houseTlYearly = compoundSeries(inp.value_of_house_tl, turkeyInflationAnnual, inp.years);
    houseTlMonthly = compoundSeries(inp.value_of_house_tl, turkeyInflationMonthly, nMonths);
  }
  const houseUsdYearly = div(houseTlYearly, dollarRatesAnnual);
  const houseUsdMonthly = div(houseTlMonthly, dollarRatesMonthly);

  // Rent
  const annualRentTlStart = inp.initial_monthly_rent_tl * 12;
  let rentTlYearly;
  if (inp.generate_random) {
    const r = randomMonthlyPath(annualRentTlStart, inp.years, inp.turkey_inflation);
    rentTlYearly = r.yearlyPath;
  } else {
    rentTlYearly = compoundSeries(annualRentTlStart, turkeyInflationAnnual, inp.years);
  }
  const rentTlMonthly = [];
  for (let i = 0; i < rentTlYearly.length - 1; i++) {
    for (let j = 0; j < 12; j++) rentTlMonthly.push(rentTlYearly[i] / 12);
  }
  rentTlMonthly.push(rentTlYearly[rentTlYearly.length - 1] / 12);

  const rentUsdYearly = div(rentTlYearly, dollarRatesAnnual);
  const rentUsdMonthly = div(rentTlMonthly, dollarRatesMonthly);
  const cumRentUsdYearly = cumsum(rentUsdYearly);
  const cumRentUsdMonthly = cumsum(rentUsdMonthly);
  const housePlusRentYearly = add(houseUsdYearly, cumRentUsdYearly);
  const housePlusRentMonthly = add(houseUsdMonthly, cumRentUsdMonthly);
  const totalCreditMinusRentYearly = sub(totalCreditUsdAnnual, cumRentUsdYearly);
  const totalCreditMinusRentMonthly = sub(totalCreditUsdMonthly, cumRentUsdMonthly);

  // Salary
  let salariesUsdYearly = null, salariesUsdMonthly = null;
  let paymentSalaryRatioYearly = null, paymentSalaryRatioMonthly = null;
  let paymentMinusRentSalaryYearly = null, paymentMinusRentSalaryMonthly = null;

  if (inp.start_salary_base > 0) {
    let sameCurrencyYearly;
    if (inp.generate_random) {
      const r = randomYearlyPath(inp.start_salary_base, inp.years, inp.salary_growth);
      sameCurrencyYearly = r.path;
    } else {
      sameCurrencyYearly = compoundSeries(inp.start_salary_base, inp.salary_growth / 100, inp.years);
    }

    if (inp.salary_currency === "EUR") salariesUsdYearly = scale(sameCurrencyYearly, inp.euro_dollar_rate);
    else if (inp.salary_currency === "USD") salariesUsdYearly = [...sameCurrencyYearly];
    else if (inp.salary_currency === "TL") salariesUsdYearly = div(sameCurrencyYearly, dollarRatesAnnual);

    salariesUsdMonthly = monthlyFromYearlySalary(
      sameCurrencyYearly, inp.months_to_increase, dollarRatesMonthly, inp.euro_dollar_rate, inp.salary_currency,
    );

    const salY = salariesUsdYearly.slice(1);
    paymentSalaryRatioYearly = annualPaymentUsd.map((p, i) => p / salY[i]);
    paymentMinusRentSalaryYearly = annualPaymentUsd.map((p, i) => (p - rentUsdYearly[i + 1]) / salY[i]);

    const salM = salariesUsdMonthly.slice(0, nMonths);
    const len = Math.min(monthlyPaymentUsd.length, salM.length);
    paymentSalaryRatioMonthly = Array.from({ length: len }, (_, i) => monthlyPaymentUsd[i] / salM[i]);
    paymentMinusRentSalaryMonthly = Array.from(
      { length: len },
      (_, i) => (monthlyPaymentUsd[i] - rentUsdMonthly[i + 1]) / salM[i],
    );
  }

  // Assets
  const assetProjections = {};
  for (const [symbol, params] of Object.entries(inp.assets || {})) {
    const { yearly, monthly } = projectAsset(params, inp);
    assetProjections[symbol] = {
      yearly, monthly,
      average_growth: params.average_growth,
      current_price: params.current_price,
    };
  }

  return {
    monthly_tl_payment: monthlyTlPayment,
    annual_tl_payment: annualTlPayment,
    total_paid_tl: monthlyTlPayment * nMonths,
    loan_amount_tl: loan,
    initial_noncredit_amount_usd: initialNoncreditUsd,
    value_of_house_usd: inp.value_of_house_tl / inp.start_dollar_tl,
    effective_dollar_growth_annual: dollarGrowthAnnual,
    effective_dollar_growth_monthly: dollarGrowthMonthly,
    effective_turkey_inflation_annual: turkeyInflationAnnual,
    effective_turkey_inflation_monthly: turkeyInflationMonthly,
    years_axis: Array.from({ length: inp.years + 1 }, (_, i) => i),
    months_axis: Array.from({ length: nMonths + 1 }, (_, i) => i),
    dollar_rates_annual: dollarRatesAnnual,
    dollar_rates_monthly: dollarRatesMonthly,
    annual_payment_usd: annualPaymentUsd,
    monthly_payment_usd: monthlyPaymentUsd,
    cumulative_payment_usd_annual: cumPaymentUsdAnnual,
    cumulative_payment_usd_monthly: cumPaymentUsdMonthly,
    total_credit_amount_usd_annual: totalCreditUsdAnnual,
    total_credit_amount_usd_monthly: totalCreditUsdMonthly,
    value_of_house_usd_yearly: houseUsdYearly,
    value_of_house_usd_monthly: houseUsdMonthly,
    rent_price_usd_yearly: rentUsdYearly,
    rent_price_usd_monthly: rentUsdMonthly,
    cumulative_rent_price_usd_yearly: cumRentUsdYearly,
    cumulative_rent_price_usd_monthly: cumRentUsdMonthly,
    house_plus_rent_yearly: housePlusRentYearly,
    house_plus_rent_monthly: housePlusRentMonthly,
    total_credit_minus_rent_usd_yearly: totalCreditMinusRentYearly,
    total_credit_minus_rent_usd_monthly: totalCreditMinusRentMonthly,
    salaries_usd_yearly: salariesUsdYearly,
    salaries_usd_monthly: salariesUsdMonthly,
    payment_salary_ratio_yearly: paymentSalaryRatioYearly,
    payment_salary_ratio_monthly: paymentSalaryRatioMonthly,
    payment_minus_rent_over_salary_yearly: paymentMinusRentSalaryYearly,
    payment_minus_rent_over_salary_monthly: paymentMinusRentSalaryMonthly,
    asset_projections: assetProjections,
  };
}
